cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/service/DraftHandler.java

```diff
@@ -0,0 +1,215 @@
+      }
+
+      SortRequest sortRequest = new SortRequest(sortBy, sortOrder);
+      responder.sendJson(draftService.listDrafts(namespace, request.getUserId(), includeConfig, sortRequest, filter));
+    });
+  }
+
+  /**
+   * Gets the details of a draft
+   */
+  @GET
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void getDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                       @PathParam("context") String namespaceName,
+                       @PathParam("draft") String draftId) {
+    respond(namespaceName, responder, (namespace) -> {
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+      responder.sendJson(draftService.getDraft(id));
+    });
+  }
+
+  /**
+   * Creates or updates a draft
+   */
+  @PUT
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void putDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                       @PathParam("context") String namespaceName,
+                       @PathParam("draft") String draftId) {
+
+    respond(namespaceName, responder, (namespace) -> {
+
+      String requestStr = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+      DraftStoreRequest<ETLConfig> draftStoreRequest = deserializeDraftStoreRequest(requestStr);
+
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+      draftService.writeDraft(id, draftStoreRequest);
+
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  /**
+   * Deletes a draft
+   */
+  @DELETE
+  @Path("v1/contexts/{context}/drafts/{draft}")
+  public void deleteDraft(HttpServiceRequest request, HttpServiceResponder responder,
+                          @PathParam("context") String namespaceName,
+                          @PathParam("draft") String draftId) {
+    respond(namespaceName, responder, (namespace) -> {
+      DraftId id = new DraftId(namespace, draftId, request.getUserId());
+
+      draftService.deleteDraft(id);
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  /**
+   * Utility method to correct deserialize the config field in the {@link DraftStoreRequest} object
+   * @param jsonStr the json string representing the DraftStoreRequest
+   * @return {@link DraftStoreRequest} object
+   */
+  private DraftStoreRequest<ETLConfig> deserializeDraftStoreRequest(String jsonStr) {
+    try {
+      DraftStoreRequest<ETLConfig> request = GSON
+        .fromJson(jsonStr, new TypeToken<DraftStoreRequest<ETLConfig>>() { }.getType());
+
+      if (request.getArtifact() == null) {
+        throw new IllegalArgumentException("artifact is null");
+      }
+
+      if (StudioUtil.isBatchPipeline(request.getArtifact())) {
+        return GSON.fromJson(jsonStr, new TypeToken<DraftStoreRequest<ETLBatchConfig>>() { }.getType());
+      } else if (StudioUtil.isStreamingPipeline(request.getArtifact())) {
```

Comment:
No need to have else after the `if` block `return`. It is cleaner to read it as:

```java
if (...) {
  return;
}
if (...) {
  return;
}
throw new IllegalArgumentException(...);
```

Reply:
done